doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,247 @@
+    called again **once**.
+
+The following use cases and behaviors should be implemented.
+
+ - The API and ReactAl should be robust this means that it should be very hard 
+    to use the API in the wrong way. Since ReactAl is working concurrently a 
+    wrong usage could lead to race conditions, extremely hard to debug bug reports 
+    (HeisenBugs) and can be frustrating to the user.
+    We list here features marked with [robust] which are only supported to 
+    increase this kind of the robustness.
+
+ - concurrently: attaching a callback at any time from anywhere. This means in particular
+    - Attaching a callback from within a callback.
+    - [robust] Updating a callback from within the same callback 
+      ```cpp
+      void onSampleReceived2(iox::popo::UntypedSubscriber & subscriber) {}
+
+      void onSampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+        myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived2);
+      }
+
+      myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived);
+      ```
+ - One can attach at most one callback to a specific event of an object.
+    - Attaching a callback to an event where a callback has been already attached overrides
+      the existing callback with the new one.
+ - concurrently: detach a callback at any time from anywhere. This means in particular
+    - Detaching a callback from within a callback.
+    - [robust] When the detach call returns we guarantee that the callback is never called
+      again. E.g. blocks till the callback is removed, if the callback is concurrently 
+      running it will block until the callback is finished and removed.
+      Exception: If a callback detaches itself it blocks until the callback is removed 
+                 and not until the callback is finished.
+    - Calling detach means that after that call the callback is no longer attached
+      even when it was not attached in the first place. Therefore it will always succeed.
+
+ - When the ReactAl goes out of scope it should detach itself from every class 
+     to which it was attached via a provided callback (like in the WaitSet).
+
+ - When the class which is attached to the ReactAl goes out of scope it should 
+    detach itself from the ReactAl via a provided callback (like in the WaitSet).
+
+### Usage Code Example
+```cpp
+ReactAl myCallbackReactal;
+iox::popo::UntypedSubscriber mySubscriber;
+
+void onSampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+  subscriber.take().and_then([&](auto & sample){
+    std::cout << "received " << std::hex << sample->payload() << std::endl;
+  });
+}
+
+void onWaitForSubscription(iox::popo::UntypedSubscriber & subscriber) {
+  std::cout << "subscribed!\n";
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived);
```

Comment:
I find this reliance on those event enums unnecessarily limiting. I am fairly sure this can be suitably generalized. In any case, those events need more explanation on a conceptual level I think.

Reply:
Restricting it to enums is making the implementation a little bit more easy. I think  YAGNI until we need a more general solution but at the moment there is not a single use case for it.